### ISSUE TITLE
Proper way to handle versioning?

### ISSUE THREAD

**Issue by xaviergonz** [mobxjs/serializr#49]
I want to add versioning, this is, the resulting JSON would have a property called _version: number and depending on this number it should instantiate the class one way or another.

What's the preferred way of doing this?

I thought about modifying the source json object itself before passing it to the deserialize method (this is, updating the old version to the latest one), but the problem comes whenever the sub-object themselves are versioned as well. 
Is there some kind of callback that I could hook into to modify the JSON of a sub-object whenever it is about to be deserialized by serializr to do this "update" transformation? 

PS: I'm using decorators at the moment, maybe there could be a class decorator that sets the factory method?

**Comment by xaviergonz** [comment]
PS: I tried this

getDefaultModelSchema(MyVersionedClass).factory = (context) => {
  // modify context.json as needed
  return new MyVersionedClass();
};

but sadly any changes made to context.json were ignored by the deserialize method later

**Comment by xaviergonz** [comment]
Actually that worked (I was changing the wrong property name). Would this be the right approach to take or is there a better one?

**Comment by alexggordon** [comment]
@xaviergonz this issue has been discussed a little bit before in the context of [polymorphism and deserialization](https://github.com/mobxjs/serializr/issues/33). 

It was decided in that example that polymorphism was too complicated to handle within serializr, however, there was a [PR](https://github.com/mobxjs/serializr/pull/30) for a bit that experimented with handling this. 

Currently, mobx-state-tree handles polymorphism through the [`union`](https://github.com/mobxjs/mobx-state-tree#utility-types) type, which is why @mweststrate probably wanted to rename `dynamicObject` in the pr to `union`. I'm wondering if @mweststrate would be open to me possibly finishing that `dynamicObject` PR, or if this is something that just shouldn't be handled in any way by serializr (outside of the `custom` type)?

**Comment by xaviergonz** [comment]
@alexggordon thanks for answering, though the question is not really about polymorphism, but rather how to address versioning, this is:

Imagine I have an object named box with a single property named 'size', and less call it Box v1.

Now suddenly I realized I need boxes that may not be squared, so I make a Box v2 that instead of 'size' uses 'width' and height', but sadly my Box v1 only has 'size'.

In C#/Java etc I'd have (hopefully) a version field as well as a custom serialization constructor that'd basically do:

if (info.GetInt32("version") == 1) { this.width = this.height = info.GetInt32("size")

In other words, at loading time the object is always "updated" to the latest version and at save time the object is always saved as the latest version available.

So my question really is: **is tweaking the json manually before passing it to serializr deserialization the best (only?) way to do it?**

The reason of all this? imagine if you had a game/app that could only load the latest version of a savegame/document and not allow you to load older ones, that could infuriate users :)

**Comment by alexggordon** [comment]
closed for inactivity
